public/app/plugins/panel/state-timeline/StateTimelinePanel.tsx

```diff
@@ -32,6 +41,69 @@ export const StateTimelinePanel: React.FC<TimelinePanelProps> = ({
     theme,
   ]);
 
+  const renderCustomTooltip = useCallback(
+    (alignedFrame: DataFrame, seriesIdx: number | null, datapointIdx: number | null) => {
+      // Not caring about multi mode in StateTimeline
+      if (seriesIdx === null && datapointIdx === null) {
+        return null;
+      }
+      const xField = alignedFrame.fields[0];
+      const xFieldFmt = xField.display || getDisplayProcessor({ field: xField, timeZone, theme });
+
+      const field = alignedFrame.fields[seriesIdx!];
+      const dataFrameFieldIndex = field.state?.origin;
+      const fieldFmt = field.display || getDisplayProcessor({ field, timeZone, theme });
+      const value = field.values.get(datapointIdx!);
+      const display = fieldFmt(value);
+      const fieldDisplayName = dataFrameFieldIndex
+        ? getFieldDisplayName(
+            data.series[dataFrameFieldIndex.frameIndex].fields[dataFrameFieldIndex.fieldIndex],
+            data.series[dataFrameFieldIndex.frameIndex],
+            data.series
+          )
+        : null;
+
+      const nextStateIdx = findNextStateIndex(field, datapointIdx!);
+      let nextStateTs;
+      if (nextStateIdx) {
+        nextStateTs = xField.values.get(nextStateIdx!);
+      }
+
+      const stateTs = xField.values.get(datapointIdx!);
+
+      let toFragment = null;
+      let durationFragment = null;
+
+      if (nextStateTs) {
+        const duration =
+          nextStateTs && formattedValueToString(toDuration(nextStateTs - stateTs, 0, Interval.Millisecond));
+        durationFragment = (
+          <>
+            <br />
+            <strong>Duration:</strong> {duration}
+          </>
+        );
+        toFragment = (
+          <>
+            {' to'} <strong>{xFieldFmt(xField.values.get(nextStateIdx!)).text}</strong>
+          </>
+        );
+      }
+
+      return (
+        <div style={{ fontSize: theme.typography.bodySmall.fontSize }}>
+          {fieldDisplayName}
+          <br />
+          <SeriesTableRow label={display.text} color={display.color || FALLBACK_COLOR} isActive />
+          From <strong>{xFieldFmt(xField.values.get(datapointIdx!)).text}</strong>
+          {toFragment}
+          {durationFragment}
+        </div>
+      );
+    },
+    [theme, timeZone, data]
+  );
```

Comment:
Custom tooltip renderer for Timeline panel - can be extracted to a separate component